File: app/src/main/java/com/yg/mykiwar/placement/draw/MainRenderer.java
package com.yg.mykiwar.placement.draw;

import android.graphics.Color;
import android.opengl.GLES20;
import android.opengl.GLSurfaceView;
import android.opengl.Matrix;

import com.google.ar.core.Frame;
import com.google.ar.core.PointCloud;
import com.google.ar.core.Session;

import java.util.ArrayList;
import java.util.List;

import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;

public class MainRenderer implements GLSurfaceView.Renderer {
    private static final String TAG = MainRenderer.class.getSimpleName();

    private boolean mViewportChanged;
    private int mViewportWidth;
    private int mViewportHeight;

    private CameraRenderer mCamera;
    private PointCloudRenderer mPointCloud;

    private List<Path> mPaths = new ArrayList<Path>();
    private List<Sphere> mSpheres = new ArrayList<Sphere>();

    private float[] mProjMatrix = new float[16];

    private RenderCallback mRenderCallback;

    public interface RenderCallback {
        void preRender();
    }

    public MainRenderer(RenderCallback callback) {
        mCamera = new CameraRenderer();
        mPointCloud = new PointCloudRenderer();

        mRenderCallback = callback;
    }

    @Override
    public void onSurfaceCreated(GL10 gl10, EGLConfig eglConfig) {
        GLES20.glEnable(GLES20.GL_DEPTH_TEST);
        GLES20.glClearColor(1.0f, 1.0f, 0.0f, 1.0f);

        mCamera.init();

        mPointCloud.init();
    }

    @Override
    public void onSurfaceChanged(GL10 gl10, int width, int height) {
        GLES20.glViewport(0, 0, width, height);
        mViewportChanged = true;
        mViewportWidth = width;
        mViewportHeight = height;
    }

    @Override
    public void onDrawFrame(GL10 gl10) {
        GLES20.glClear(GLES20.GL_COLOR_BUFFER_BIT | GLES20.GL_DEPTH_BUFFER_BIT);

        mRenderCallback.preRender();

        GLES20.glDepthMask(false);
        mCamera.draw();
        GLES20.glDepthMask(true);

        mPointCloud.draw();

        for (int i = 0; i < mSpheres.size(); i++) {
            Sphere sphere = mSpheres.get(i);
            if (!sphere.isInitialized()) {
                sphere.init();
            }
            sphere.draw();
        }

        for (int i = 0; i < mPaths.size(); i++) {
            Path path = mPaths.get(i);
            if (!path.isInitialized()) {
                path.init();
            }
            path.update();
            path.draw();
        }
    }

    public int getTextureId() {
        return mCamera == null ? -1 : mCamera.getTextureId();
    }

    public void onDisplayChanged() {
        mViewportChanged = true;
    }

    public boolean isViewportChanged() {
        return mViewportChanged;
    }

    public int getWidth() {
        return mViewportWidth;
    }

    public int getHeight() {
        return mViewportHeight;
    }

    public void updateSession(Session session, int displayRotation) {
        if (mViewportChanged) {
            session.setDisplayGeometry(displayRotation, mViewportWidth, mViewportHeight);
            mViewportChanged = false;
        }
    }

    public void transformDisplayGeometry(Frame frame) {
        mCamera.transformDisplayGeometry(frame);
    }

    public void updatePointCloud(PointCloud pointCloud) {
        mPointCloud.update(pointCloud);
    }

    public void setProjectionMatrix(float[] matrix) {
        System.arraycopy(matrix, 0, mProjMatrix, 0, 16);

        mPointCloud.setProjectionMatrix(matrix);
    }

    public void updateViewMatrix(float[] matrix) {
        mPointCloud.setViewMatrix(matrix);

        for (int i = 0; i < mSpheres.size(); i++) {
            mSpheres.get(i).setViewMatrix(matrix);
        }

        for (int i = 0; i < mPaths.size(); i++) {
            mPaths.get(i).setViewMatrix(matrix);
        }
    }

    public void setModelMatrix(float[] matrix) {
    }

    public void addPath(float x, float y, float z) {
        Path currentPath = new Path(50, Color.WHITE);
        currentPath.setProjectionMatrix(mProjMatrix);

        float[] identity = new float[16];
        Matrix.setIdentityM(identity, 0);
        currentPath.setModelMatrix(identity);

        currentPath.updatePoint(x, y, z);

        mPaths.add(currentPath);
    }

    public void addPoint(float x, float y, float z) {
        if (mPaths.size() >= 1) {
            Sphere currentPoint = new Sphere(0.01f, Color.GREEN);
            currentPoint.setProjectionMatrix(mProjMatrix);

            float[] translation = new float[16];
            Matrix.setIdentityM(translation, 0);
            Matrix.translateM(translation, 0, x, y, z);
            currentPoint.setModelMatrix(translation);

            Path path = mPaths.get(mPaths.size() - 1);
            path.updatePoint(x, y, z);
        }
    }

    public void removePath() {
        if (mPaths.size() >= 1) {
            mPaths.remove(mPaths.size() - 1);
        }
    }

 }

File: app/src/main/java/com/yg/mykiwar/temp/MainRenderer.kt
package com.yg.mykiwar.temp

import android.content.Context
import android.graphics.Color
import android.opengl.GLES20
import android.opengl.GLSurfaceView
import com.google.ar.core.Frame
import com.google.ar.core.Plane
import com.google.ar.core.PointCloud
import com.google.ar.core.Session

import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10

class MainRenderer(var context : Context, var callback : RenderCallback) : GLSurfaceView.Renderer {

    private val TAG = MainRenderer::class.java.simpleName

    private var mViewportChanged: Boolean = false
    private var mViewportWidth: Int = 0
    private var mViewportHeight: Int = 0

    private val mCamera: CameraRenderer
    private val mPointCloud: PointCloudRendererT
    private val mPlane: PlaneRendererT

    private val mObj: ObjectRendererT

    private val mRenderCallback: RenderCallback

    interface RenderCallback {
        fun preRender()
    }

    init {
        mCamera = CameraRenderer()

        mPointCloud = PointCloudRendererT()

        mPlane = PlaneRendererT(Color.GRAY, 0.5f)

        mObj = ObjectRendererT(context, "models/andy.obj", "models/andy.png")

        mRenderCallback = callback
    }


    override fun onDrawFrame(gl: GL10?) {
        GLES20.glClear(GLES20.GL_COLOR_BUFFER_BIT or GLES20.GL_DEPTH_BUFFER_BIT)

        mRenderCallback.preRender()

        GLES20.glDepthMask(false)
        mCamera.draw()
        GLES20.glDepthMask(true)

        mPointCloud.draw()

        mPlane.draw()

        mObj.draw()
    }

    override fun onSurfaceChanged(gl: GL10?, width: Int, height: Int) {
        GLES20.glViewport(0, 0, width, height)
        mViewportChanged = true
        mViewportWidth = width
        mViewportHeight = height
    }

    override fun onSurfaceCreated(gl: GL10?, config: EGLConfig?) {
        GLES20.glEnable(GLES20.GL_DEPTH_TEST)
        GLES20.glEnable(GLES20.GL_BLEND)
        GLES20.glBlendFunc(GLES20.GL_SRC_ALPHA, GLES20.GL_ONE_MINUS_SRC_ALPHA)
        GLES20.glClearColor(1.0f, 1.0f, 0.0f, 1.0f)

        mCamera.init()

        mPointCloud.init()

        mPlane.init()

        mObj.init()
    }

    fun getTextureId(): Int {
        return mCamera?.getTextureId() ?: -1
    }

    fun onDisplayChanged() {
        mViewportChanged = true
    }

    fun isViewportChanged(): Boolean {
        return mViewportChanged
    }

    fun getWidth(): Int {
        return mViewportWidth
    }

    fun getHeight(): Int {
        return mViewportHeight
    }

    fun updateSession(session: Session, displayRotation: Int) {
        if (mViewportChanged) {
            session.setDisplayGeometry(displayRotation, mViewportWidth, mViewportHeight)
            mViewportChanged = false
        }
    }

    fun transformDisplayGeometry(frame: Frame) {
        mCamera.transformDisplayGeometry(frame)
    }

    fun updatePointCloud(pointCloud: PointCloud) {
        mPointCloud.update(pointCloud)
    }

    fun updatePlane(plane: Plane) {
        mPlane.update(plane)
    }


    fun setObjModelMatrix(matrix: FloatArray) {
        mObj.setModelMatrix(matrix)
    }

    fun setProjectionMatrix(matrix: FloatArray) {
        mPointCloud.setProjectionMatrix(matrix)
        mPlane.setProjectionMatrix(matrix)
        mObj.setProjectionMatrix(matrix)
    }

    fun updateViewMatrix(matrix: FloatArray) {
        mPointCloud.setViewMatrix(matrix)
        mPlane.setViewMatrix(matrix)
        mObj.setViewMatrix(matrix)
    }

    fun getObjMinPoint(): FloatArray {
        return mObj.getMinPoint()
    }

    fun getObjMaxPoint(): FloatArray {
        return mObj.getMaxPoint()
    }


}